### PR TITLE
Upgrade to Rails 6.0

### PR DESCRIPTION
**Pull request by etagwerker**
Hey @rhymes @tannakartikey,

I went ahead and started this exploratory PR. 

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR adds a couple of changes to the project: 

- Adds `Gemfile.next` to make it easier to dual boot the app 
- Adds one gemfile to the Travis CI configuration

- [ ] Sorry about the Rubocop TODO changes, but it wouldn't let me commit otherwise. I'll try to remove them later. If you are curious, it was complaining about a duplicate `gem "rails"` in the Gemfile. 😺 

## Related Tickets & Documents

https://github.com/thepracticaldev/dev.to/issues/2879

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

N/A 

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I'm not sure this PR needs them? 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?

I'm pretty sure you cookies generated by Rails 5.2 are not going to be compatible with Rails 6.0. So users might get signed out? (not sure) 

## [optional] What gif best describes this PR or how it makes you feel?

Both excited and freaked out! xD

giphy

### CODE REVIEW

**Comment by rhymes** [comment]
Look great! I think we can ignore rubocop for now with `git commit --no-verify`.

I think we should eventually run `rails app:update` as indicated in the [upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task) though by doing that we'd necessarily change files that are shared by both.

Anyway, I think that this multi step strategy makes sense:

- this PR activates CI on Rails 6
- we fix deprecations and errors
- another PR actually upgrades the code to use Rails 6 and switches `Gemfile.next` to Rails master

What do you think @etagwerker ?

**Comment by etagwerker** [comment]
@rhymes Sounds like a good plan. 

I think this PR can be a guide for the upgrade. Any changes that are backwards compatible, should be submitted as a PR targeting `master`.

**Comment by rhymes** [comment]
@etagwerker both me and @maestromac tried to push changes to this PR but we can't:

```shell
remote: Permission to fastruby/dev.to.git denied to rhymes.
fatal: unable to access 'https://github.com/fastruby/dev.to.git/': The requested URL returned error: 403
```

is it a protected fork?

**Comment by etagwerker** [comment]
@rhymes @maestromac Not sure why that happened. Maybe I checked the wrong box when I submitted the PR (?) 

Just sent you invitations to the repo which should fix it: https://github.com/fastruby/dev.to/invitations

**Comment by rhymes** [comment]
@etagwerker it works, thanks!

**Comment by rhymes** [comment]
@etagwerker sorry I had to basically undo the `Gemfile.next` trick as it's wasn't working both locally and on Travis - https://travis-ci.com/github/thepracticaldev/dev.to/jobs/338821108.

I kept encountering various failures in tests (running them with `bundle exec --gemfile=Gemfile.next`) and after booting the app with that Gemfile I couldn't login or logout.

After spending sometime in the debugger I realized there was an easier way to figure out if the app was or wasn't working with Rails 6: actually update it to it.

So I did, and as you can see here https://travis-ci.com/github/thepracticaldev/dev.to/builds/167789826, all tests were green. Also locally everything was working.

I still think the `Gemfile.next` trick is interesting but maybe we can circle back on that after this upgrade so that we can track Rails 6.1 effectively. One thing that I foresee is going to be hard to work with is how quickly stale `Gemfile.next.lock` becomes. Yesterday dependabot sent us a bunch of updates and I had to manually update it with `bundle update --gemfile=Gemfile.next` and still I had slightly different versions of gems that weren't related to Rails (that's because we don't lock specific versions for most of the gems and the Bundle gemfile resolution algorithm).

Long story short: after fixing a bunch of visible deprecations, I think we have DEV running on Rails 6 and it was pretty easy 🔥 Still a few kinks but we're almost there.

Thanks for the help and the initial nudge. As I said I plan to circle back on `Gemfile.next` with a branch to track 6.1 once all the dust is settled.

Next steps are all about "Rails-six-fying" the code by at least migrating to Zeitwerk as the auto loader and enabling parallel testing if possible. We'll do that throughout the following weeks, let me know if you want to help :)

File: app/controllers/chat_channel_memberships_controller.rb [hunk 1]
```diff
@@ -117,7 +117,9 @@ def respond_to_invitation(previous_status)
         flash[:settings_notice] = "Invitation to  #{channel_name} accepted. It may take a moment to show up in your list."
       else
         send_chat_action_message("@#{current_user.username} added @#{@chat_channel_membership.user.username}", current_user, @chat_channel_membership.chat_channel_id, "joined")
-        NotifyMailer.channel_invite_email(@chat_channel_membership, @chat_channel_membership.user).deliver_later
+
+        NotifyMailer.with(membership: @chat_channel_membership, inviter: @chat_channel_membership.user).channel_invite_email.deliver_later
```
Comment (rhymes) [inline app/controllers/chat_channel_memberships_controller.rb:121]: this is the new API in Rails 6 for parameterized mailers: https://github.com/rails/rails/pull/34591

File: app/labor/sticky_article_collection.rb [hunk 2]
```diff
@@ -24,7 +24,7 @@ def tag_articles
     @tag_articles ||= Article.published.tagged_with(article_tags, any: true).
       includes(:user).
       where("positive_reactions_count > ? OR comments_count > ?", reaction_count_num, comment_count_num).
-      where.not(id: article.id, user_id: article.user_id).
+      where.not(id: article.id).where.not(user_id: article.user_id).
```
Comment (rhymes) [inline app/labor/sticky_article_collection.rb:27]: Rails 6.1 is going to change the algorithm from NOR to NAND if you pass multiple arguments to a `where.not`, so I fixed the deprecation, see https://github.com/rails/rails/blob/6-0-stable/activerecord/CHANGELOG.md#rails-600rc1-april-24-2019

File: app/services/audit/event/util.rb [hunk 3]
```diff
@@ -1,3 +1,4 @@
+# NOTE: [Rails 6] Util class is obsolete in Rails 6
```
Comment (rhymes) [inline app/services/audit/event/util.rb:1]: Going to leave it there for now, I'll slowly chip out other deprecated stuff and aim to upgrade to 6.0 defaults in the future

File: bin/setup [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env ruby
```
Comment (rhymes) [inline bin/setup:1]: ran `rails app:update` here

File: bin/update [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env ruby
```
Comment (rhymes) [inline bin/update:1]: ran `rails app:update` here. We could probably remove this but I want to look into the new `rails db:*` tasks before, I think there are some that can simplify our setup

File: config/application.rb [hunk 6]
```diff
@@ -50,24 +50,9 @@ class Application < Rails::Application
     # Therefore we disable "per_form_csrf_tokens" for the time being.
     config.action_controller.per_form_csrf_tokens = false
 
-    # Enable CORS for API v0
-    # (logging is only activated when debug is enabled)
-    debug_cors = ENV["DEBUG_CORS"].present? ? true : false
```
Comment (rhymes) [inline config/application.rb:55]: moved in the default CORS initializer Rails 6 creates

File: config/application.rb [hunk 7]
```diff
@@ -50,24 +50,9 @@ class Application < Rails::Application
...
-          # allow read operations, disallow custom headers (eg. api-key) and enable preflight caching
-          # NOTE: Chrome caps preflight caching at 2 hours, Firefox at 24 hours
-          # see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age#Directives
-          resource "/api/#{resource_name}/*", methods: %i[head get options], headers: [], max_age: 2.hours.to_i
-        end
-      end
-    end
+    # NOTE: [Rails 6]
+    # To improve security, Rails embeds the purpose and expiry metadata inside encrypted or signed cookies value.
+    config.action_dispatch.use_cookies_with_metadata = false
```
Comment (rhymes) [inline config/application.rb:55]: this won't invalidate existing cookies but I'd rather activate it separately, see https://github.com/rails/rails/pull/32937

File: config/boot.rb [hunk 8]
```diff
@@ -1,5 +1,4 @@
 ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../Gemfile", __dir__)
```
Comment (rhymes) [inline config/boot.rb:1]: ran `rails app:update` here

File: config/environments/production.rb [hunk 9]
```diff
@@ -1,7 +1,4 @@
 Rails.application.configure do
-  # Verifies that versions and hashed value of the package contents in the project's package.json
-  config.webpacker.check_yarn_integrity = false
```
Comment (rhymes) [inline config/environments/production.rb:3]: this was long deprecated and removed. See #7478

File: spec/models/organization_spec.rb [hunk 10]
```diff
@@ -240,14 +240,22 @@
 
       it "updates the paths of the organization's articles" do
         new_slug = "slug_#{rand(10_000)}"
-        organization.update(slug: new_slug)
+
+        sidekiq_perform_enqueued_jobs(only: Organizations::UpdateOrganizationArticlesPathsWorker) do
```
Comment (rhymes) [inline spec/models/organization_spec.rb:244]: no idea why this is currently working in master as it shouldn't since it didn't execute the jobs needed for the to pass 🤷

File: spec/requests/api/v0/articles_spec.rb [hunk 11]
```diff
@@ -45,7 +45,7 @@
     context "without params" do
       it "returns json response" do
         get api_articles_path
-        expect(response.content_type).to eq("application/json")
+        expect(response.media_type).to eq("application/json")
```
Comment (rhymes) [inline spec/requests/api/v0/articles_spec.rb:48]: `content_type` will contain the full content type (with encoding and other options) in Rails 6.1, so 6.0 signals this access with a deprecation and suggests to move to `media_type` to check only that

**Comment by rhymes** [comment]
I would like to merge https://github.com/thepracticaldev/dev.to/pull/8004 ahead of this and then rebase this one on top of master

**Comment by rhymes** [comment]
Also, not sure why these https://travis-ci.com/github/thepracticaldev/dev.to/builds/167890168#L1327-L1330 are failing 🤷

**Comment by etagwerker** [comment]
@rhymes Thanks for the update re: dual booting! Do you want me to submit an issue to keep track of this idea? It can be revisited once you get Rails 6.0 up and running in production. 😄 

Regarding the failures in `spec/system/notifications/notifications_page_spec.rb` I'm not sure what is the deal there. I tried to debug it a bit and it looks like a timing issue (?) -- I don't intend for this change to make it to `master`, but I wanted to try it in Travis CI: https://github.com/thepracticaldev/dev.to/pull/7658/commits/51630de2eb054994a785475281aa7bbbe738af4c 

One thing that is weird is that I see that spec working fine if I drop the `headless` option for the driver: 

```
Capybara.register_driver :headless_chrome do |app|
  options = Selenium::WebDriver::Chrome::Options.new(
    args: %w[no-sandbox disable-gpu window-size=1920,1080 --enable-features=NetworkService,NetworkServiceInProcess],
    log_level: :error,
  )

  Capybara::Selenium::Driver.new app, browser: :chrome, options: options
end
```

So I think it is a failure related to JS + running the scenario with a headless driver. Maybe some scenarios require to run with a non-headless driver?

**Comment by rhymes** [comment]
@etagwerker 

> @rhymes Thanks for the update re: dual booting! Do you want me to submit an issue to keep track of this idea? It can be revisited once you get Rails 6.0 up and running in production. 😄

yeah, sure! We could also use it for a discussion around that as we have also to figure out what impact having double the amount of builds has

> Regarding the failures in `spec/system/notifications/notifications_page_spec.rb` I'm not sure what is the deal there. I tried to debug it a bit and it looks like a timing issue (?) -- I don't intend for this change to make it to `master`, but I wanted to try it in Travis CI: [51630de](https://github.com/thepracticaldev/dev.to/commit/51630de2eb054994a785475281aa7bbbe738af4c)

We do have a `wait_for_javascript` helper here

https://github.com/thepracticaldev/dev.to/blob/9c58ebc61e3fd581da3f8bf83c9904d162cb03e9/spec/support/initializers/capybara.rb#L38

I'll see if I can replace the `sleep` with that. Unfortunately we have several flaky tests and our code base is dependent on order, introducing a new version of the framework is bound to surface some of these problems or generate new ones.

I'll also ping other team members that are more familiar with such notifications code. Thanks for the effort!

> One thing that is weird is that I see that spec working fine if I drop the `headless` option for the driver:

That is weird indeed.

> So I think it is a failure related to JS + running the scenario with a headless driver. Maybe some scenarios require to run with a non-headless driver?

I always thought there was no difference, except the Chrome window appearing or not. I wonder also if it's a question of Chrome version. We do test against `chrome: 'stable'` in `.travis.yml` but Chrome stable changes in time...

**Comment by rhymes** [comment]
@etagwerker for example, locally I have several of `./spec/system/notifications/notifications_page_spec.rb` tests fail with:

```shell
     3.2) Failure/Error: raise ActionController::RoutingError, "No route matches [#{env['REQUEST_METHOD']}] #{env['PATH_INFO'].inspect}"

          ActionController::RoutingError:
            No route matches [GET] "/stories/warm_comments/notifications"
```

and they fail both with or without `headless`

File: app/controllers/chat_channel_memberships_controller.rb (hunk 1 above)
Comment (maestromac) [reply app/controllers/chat_channel_memberships_controller.rb:121]: Oh this is super neat!

Comment (vaidehijoshi) [reply app/controllers/chat_channel_memberships_controller.rb:121]: Whoa! Gotta teach myself to start thinking in the Rails 6 way 😅

File: app/queries/internal/moderators_query.rb [hunk 12]
```diff
@@ -14,7 +14,7 @@ def self.call(relation: User.all, options: {})
                      role_id_for(:trusted),
                    ).order("users.comments_count DESC")
                  else
-                   relation.joins(:users_roles).
+                   relation.joins(:roles).
```
Comment (vaidehijoshi) [inline app/queries/internal/moderators_query.rb:17]: Is this change due to a deprecation?

Comment (rhymes) [reply app/queries/internal/moderators_query.rb:17]: Nope, it just doesn't work without that change. Not totally sure why, though it works in current master as well if I change it to `:roles` 🤷

File: .travis.yml [hunk 13]
```diff
@@ -15,6 +15,8 @@ addons:
       - $(ls tmp/screenshots/*.png | tr "\n" ":")
       - $(ls tmp/csvs/*.csv | tr "\n" ":")
     debug: true
+gemfile:
+- Gemfile
```
Comment (etagwerker) [inline .travis.yml:19]: @rhymes I guess now that we are not using dual booting, this change is unnecessary. 😄

File: config/application.rb [hunk 14]
```diff
@@ -20,7 +20,7 @@
 module PracticalDeveloper
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
-    config.load_defaults 5.1
+    config.load_defaults 5.1 # NOTE: [Rails 6] we should at least work towards updating this to 5.2
```
Comment (etagwerker) [inline config/application.rb:23]: @rhymes Does it cause a lot of problems if we try with 5.2 or 6.0?

File: lib/tasks/app_setup.rake [hunk 15]
```diff
@@ -5,12 +5,7 @@ namespace :app_initializer do
     Rake::Task["search:setup"].execute
 
     puts "\n== Preparing database =="
-    begin
-      Rake::Task["db:migrate"].execute
-    rescue ActiveRecord::NoDatabaseError
-      puts "\n== Creating and Seeding database =="
-      system("bin/rails db:setup")
-    end
+    system!('bin/rails db:prepare')
```
Comment (mstruve) [inline lib/tasks/app_setup.rake:8]: 🔥 🔥 🔥

File: app/queries/internal/moderators_query.rb (hunk 12 above)
Comment (citizen428) [reply app/queries/internal/moderators_query.rb:17]: `roles` is the actual association table and `users_roles` the join table, so I think it actually makes more sense like this anyway.

File: config/initializers/cors.rb [hunk 16]
```diff
@@ -0,0 +1,19 @@
+# Be sure to restart your server when you modify this file.
+
+# Enable CORS for API v0 (logging is only activated when debug is enabled)
+debug_cors = ENV["DEBUG_CORS"].present? ? true : false
```
Comment (citizen428) [inline config/initializers/cors.rb:4]: Did Rails add it like this? `present?` already returns a boolean, so no need for the ternary.

File: .travis.yml (hunk 13 above)
Comment (rhymes) [reply .travis.yml:19]: yeah, I kept it as a reminder :D

File: config/application.rb (hunk 14 above)
Comment (rhymes) [reply config/application.rb:23]: Yep. if you look at https://github.com/thepracticaldev/dev.to/blob/master/config/initializers/new_framework_defaults_5_2.rb we still have at least one setting disabled. I think we can bump it later on making sure that setting stays disabled for the time being and them bump to 5.2.

I wouldn't bump to 6.0 yet, before we check each setting.

Anyway, I plan to send follow up PRs to address all the remaining deprecations and possibly "Rails-sixify" the code :D

**Comment by rhymes** [comment]
@mstruve 

> This looks SUPER freaking solid! Question. We have some AR relations that are cached in Redis, do we know if those are going to break with this upgrade? For example:

Great question! How are they serialized now in the cache? Does Rails call `Marshal.dump()` on the objects?

If so, I did a super layman check (loaded the same object in Rails 5.2 and Rails 6 and checked if the strings were the same) and I can tell you they are not. So yeah, the marhsalling changes.

**BUT**

I did the following and it worked:

```ruby
# with Rails 6
> File.open('user.dump', 'wb') { |f| f.write(Marshal.dump(user)) }
# then with Rails 5.2
[1] pry(main)> user = Marshal.load(File.read('user.dump'))
=> #<User id: 11, articles_count: 0, available_for: nil, badge_achievements_count: 0, behance_url: nil, bg_color_hex: nil, blocked_by_count: 0, blocking_others_count: 0, checked_co...

# then with Rails 5.2
[4] pry(main)> File.open('user2.dump', 'wb') { |f| f.write(Marshal.dump(user2)) }

# and reloaded with Rails 6
[1] pry(main)> user2 = Marshal.load(File.read('user2.dump'))
=> #<User id: 11, articles_count: 0, available_for: nil, badge_achievements_count: 0, behance_url: nil, bg_color_hex: nil, blocked_by_count: 0, blocking_others_count: 0, checked_...
```

so the format is compatible :)

File: config/initializers/cors.rb (hunk 16 above)
Comment (rhymes) [reply config/initializers/cors.rb:4]: Nope, I just copy pasted from our code from 

https://github.com/thepracticaldev/dev.to/blob/5e3f645d299f4e583b11c9322cebc14b414404c3/config/application.rb#L55

:D 

I'll fix it eventually!

File: Gemfile [hunk 17]
```diff
@@ -101,7 +101,7 @@ gem "storext", "~> 3.1" # Add type-casting and other features on top of ActiveRe
 gem "stripe", "~> 5.22" # Ruby library for the Stripe API
 gem "strong_migrations", "~> 0.6" # Catch unsafe migrations
 gem "timber", "~> 3.0" # Great Ruby logging made easy
-gem "timber-rails", "~> 1.0" #  Timber integration for Rails
+gem "timber-rails", github: "timberio/timber-ruby-rails", branch: "master" # Timber integration for Rails
```
[the remark this answers is not in the document]
Comment (rhymes) [reply Gemfile:104]: @joshpuetz subcribed to the issue. Hopefully we can get rid of the gem altogether (I know it's in the SRE team's plans)

File: config/initializers/new_framework_defaults_6_0.rb [hunk 18]
```diff
@@ -0,0 +1,45 @@
+# Be sure to restart your server when you modify this file.
```
[the remark this answers is not in the document]
Comment (rhymes) [reply config/initializers/new_framework_defaults_6_0.rb:1]: yeah, this is installed by default by `rails app:update`. I'm going to flip them on (the ones we can enable at least), one at at a time, the same way we did for rails 5.2

See #4216, #4829, #4826, #4827, #5005, #5057 and #5329 :D

**Comment by etagwerker** [comment]
@rhymes @joshpuetz awesome, thanks! 💯

**Comment by rhymes** [comment]
@etagwerker thank you!
